src/cpp/session/include/session/SessionProjectTemplate.hpp

```diff
@@ -0,0 +1,261 @@
+/*
+ * SessionProjectTemplate.hpp
+ *
+ * Copyright (C) 2009-12 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#ifndef SESSION_PROJECT_TEMPLATE_HPP
+#define SESSION_PROJECT_TEMPLATE_HPP
+
+#include <core/Base64.hpp>
+#include <core/FileSerializer.hpp>
+#include <core/StringUtils.hpp>
+#include <core/json/Json.hpp>
+#include <core/json/JsonRpc.hpp>
+#include <core/text/CsvParser.hpp>
+
+#include <boost/system/error_code.hpp>
+#include <boost/foreach.hpp>
+
+namespace rstudio {
+namespace core {
+   class Error;
+}
+}
+
+namespace rstudio {
+namespace session {
+namespace modules {
+namespace projects {
+namespace templates {
+
+#define kProjectTemplateWidgetTypeCheckBox  "checkbox"
+#define kProjectTemplateWidgetTypeSelectBox "selectbox"
+#define kProjectTemplateWidgetTypeTextInput "textinput"
+#define kProjectTemplateWidgetTypeFileInput "fileinput"
+
+static const char* const kWidgetTypes[] = {
+   kProjectTemplateWidgetTypeCheckBox,
+   kProjectTemplateWidgetTypeSelectBox,
+   kProjectTemplateWidgetTypeTextInput,
+   kProjectTemplateWidgetTypeFileInput
+};
+
+struct ProjectTemplateWidgetDescription
+{
+   std::string parameter;
+   std::string type;
+   std::string label;
+   std::vector<std::string> fields;
+
+   core::json::Value toJson() const
+   {
+      core::json::Object object;
+
+      object["parameter"] = parameter;
+      object["type"]      = type;
+      object["label"]     = label;
+
+      object["fields"]    = core::json::toJsonArray(fields);
+
+      return object;
+   }
+
+   static ProjectTemplateWidgetDescription fromJson(core::json::Object& object)
+   {
+      ProjectTemplateWidgetDescription ptwd;
+      
+      core::Error error = core::json::readObject(
+               object,
+               "parameter", &ptwd.parameter,
+               "type",      &ptwd.type,
+               "label",     &ptwd.label);
+      
+      if (error)
+         LOG_ERROR(error);
```

Comment:
I think it'd be better here to return the error (and take a pointer to a ptwd as a parameter) so that the caller knows not to process the template if it's malformed.